source/gui/settingsDialogs.py

```diff
@@ -5443,3 +5523,197 @@ def onFilterEditTextChange(self, evt):
+				)
+				if self._testStatus == SetURLDialog._URLTestStatus.FAILED
+				else _(
+					# Translators: Message shown to users when saving an untested URL to NVDA's settings.
+					"The URL you have entered has not been tested. Are you sure you want to save it without attempting to connect to it first?",
+				),
+				# Translators: The title of a dialog.
+				_("Warning"),
+				wx.OK | wx.CANCEL | wx.CANCEL_DEFAULT | wx.ICON_WARNING,
+				self,
+			)
+			if ret == wx.OK:
+				shouldSave = True
+			else:
+				return
+		else:
+			shouldSave = True
+
+		if shouldSave:
+			self._saveToConfig()
+		super().onOk(evt)
+
+	def _onTextChange(self, evt: wx.CommandEvent):
+		"""Enable the "Test..." button only when there is text in the URL control, and change the URL's test status when the URL changes."""
+		value = self._url
+		self._testButton.Enable(not (len(value) == 0 or value.isspace()))
+		self._testStatus = SetURLDialog._URLTestStatus.UNTESTED
+
+	def _onTest(self, evt: wx.CommandEvent):
+		"""Normalize the URL, start a background thread to test it, and show an indeterminate progress dialog to the user."""
+		self._normalize()
+		t = threading.Thread(
+			name=f"{self.__class__.__module__}.{self._onTest.__qualname__}",
+			target=self._bg,
+			daemon=True,
+		)
+		self._progressDialog = gui.IndeterminateProgressDialog(
+			self,
+			title=self.title,
+			# Translators: A message shown to users when connecting to a URL to ensure it is valid.
+			message=_("Validating URL..."),
+		)
+		t.start()
+
+	def _bg(self):
+		"""Background URL connection thread."""
+		try:
+			with requests.get(self._urlTransformer(self._url)) as r:
+				r.raise_for_status()
+			self._success()
+		except RequestException as e:
+			log.debug(f"Failed to check URL: {e}")
+			self._failure(e)
+
+	def _success(self):
+		"""Notify the user that we successfully connected to their URL."""
+		self._testStatus = SetURLDialog._URLTestStatus.PASSED
+		self._cleanUp()
+		wx.CallAfter(
+			gui.messageBox,
+			# Translators: Message shown to users when testing a given URL has succeeded.
+			_("Successfully connected to the given URL."),
+			# Translators: The title of a dialog presented when a test succeeds
+			_("Success"),
+			wx.OK,
+		)
+
+	def _failure(self, error: Exception):
+		"""Notify the user that testing their URL failed."""
+		self._testStatus = SetURLDialog._URLTestStatus.FAILED
+		self._cleanUp()
+		wx.CallAfter(
+			gui.messageBox,
+			_(
+				# Translators: Message displayed to users when testing a URL has failed.
+				"Unable to connect to the given URL. Check that you are connected to the internet and the URL is correct.",
+			),
+			# Translators: The title of a dialog presented when an error occurs.
+			"Error",
+			wx.OK | wx.ICON_ERROR,
+		)
+
+	def _cleanUp(self):
+		"""Clean up after testing a URL."""
+		wx.CallAfter(self._progressDialog.done)
+		self._progressDialog = None
```

Comment:
I am unsure if this is safe to be in a helper function and called directly from other background thread functions. Thoughts, @seanbudd?